### PR TITLE
Update ECP-authenticated HTTPS requests

### LINKED ISSUE #3995: Update ECP authenticated pulls
@duncanmmacleod says we should replace the https code in https://github.com/gwastro/pycbc/blob/bc45bb13f3ff08d00de5e977e6b7b0a4fc61ff2f/pycbc/workflow/configuration.py#L199 with something like:

```from ciecplib import Session
with Session(**kwargs):
    sess.get(...)```
    
  ```actually, if its a one-shot, you can probably just use ciecplib.get and ignore the session entirely```
    
 Ian to look at this unless someone beats him to it. See also [ciecplib.readthedocs.io](http://ciecplib.readthedocs.io/).

### PR DESCRIPTION
**Pull request by duncanmmacleod**
This PR closes #3995 by reimplmenting the guts of `pycbc.workflow.resolve_url` with a calls to `ciecplib`, which handles ECP authentication.

### CODE REVIEW

**Comment by duncanmmacleod** [comment]
@ahnitz, `ciecplib` is _not_ LIGO specific, however, the idea of accessing proprietary information probably is. But, I'm not really a pycbc team member, so am happy to be told where to put this requirement.

The fundamental problem of on-the-fly compilation of requirements is definitely not anything to do with LIGO, but is principally because pypi.python.org is the wild west in terms of uploading pre-compiled binaries. *cough* use conda *cough*

**Comment by ahnitz** [comment]
> @ahnitz, `ciecplib` is _not_ LIGO specific, however, the idea of accessing proprietary information probably is. But, I'm not really a pycbc team member, so am happy to be told where to put this requirement.
> 
You submitted a PR. Welcome to the team! There's not exactly membership requirements, it's a do-ocracy and you've done something. BTW, it's always really nice to see a PR that removes a ton of code to replace with much simpler logic. 

> The fundamental problem of on-the-fly compilation of requirements is definitely not anything to do with LIGO, but is principally because pypi.python.org is the wild west in terms of uploading pre-compiled binaries. _cough_ use conda _cough_

Conda certainly has some use cases, but the python ecosystem more broadly is built around pypi. I'd like to ensure we always support a pypi centric model (in addition to others as people want) and are working within the broader scientific / astronomy python ecosystem. That means  I always want to make sure it is easy to install pycbc from pypi without too many 'hidden' system requirements.

**Comment by titodalcanton** [comment]
I do not immediately see any downsides in moving dependencies to`requirements-ldg.txt`, I just note that it should be described in the install instructions.

**Comment by josh-willis** [comment]
@duncanmmacleod Should this in fact be `requirements-igwn.txt`? Or at least not something LIGO specific?

**Comment by duncanmmacleod** [comment]
> @duncanmmacleod Should this in fact be `requirements-igwn.txt`? Or at least not something LIGO specific?

Yes, probably, nothing in this is inherently LIGO-specific. This workflow _should_ work for KAGRA members to use their own native credentials to work with PyCBC for IGWN.

**Comment by ahnitz** [comment]
@duncanmmacleod It doesn't seem like there is any further discussion here, so assuming consent, can you move the new dependency to a requirements-iwgn.txt and update the relevant install instructions that @titodalcanton is pointing to?

**Comment by duncanmmacleod** [comment]
> @duncanmmacleod It doesn't seem like there is any further discussion here, so assuming consent, can you move the new dependency to a requirements-iwgn.txt and update the relevant install instructions that @titodalcanton is pointing to?

It is appropriate to add a setuptools `extra`, rather than a requirements.txt file, so that users can `pip install pycbc[igwn]`?

**Comment by ahnitz** [comment]
@duncanmmacleod I think you'd want both. requirements allows specific versions which might be useful for the docker / iwgn releases, while setup.py should just be the minimum requirements for the functionality. 

I'm not the holdup on this PR am I?

**Comment by titodalcanton** [comment]
@duncanmmacleod I think the IGWN or LDG specific instructions should go in this section:

https://pycbc.org/pycbc/latest/html/install.html#development-build-on-ldg-igwn-clusters

Apart from this, I have no other comments on the PR as is.

**Comment by duncanmmacleod** [comment]
@spxiwh, I've just updated this PR with a new `igwn` extra, include some docs updates, but we still have the issue of the myriad test failures. Should I hack up more of the CI to have those pass?

**Comment by duncanmmacleod** [comment]
> > @spxiwh, I've just updated this PR with a new `igwn` extra, include some docs updates, but we still have the issue of the myriad test failures. Should I hack up more of the CI to have those pass?
> 
> There shouldn't be failures for most tests, I think because the dependency should now be optional and not in the requirements, right?

The new `ciecplib` requirement is now ___not___ in the setup.py `install_requires` list, if that's what you mean.

File: requirements.txt
Comment (ahnitz) [inline requirements.txt:35]: Ah, this will be pulled in from here for most of the tests. I think if you move this to a requirements-igwn.txt and then install that also for the docker image (so that it gets pulled in for singularity I think by IGWN?), then the tests should mostly pass. I suspect a few other changes may be needed for the docker image, but not globally. 

@spxiwh Am I forgetting anything you all use that will need these?

**Comment by duncanmmacleod** [comment]
Ok, I have now stepped back to not modify any CI, I am confused as to what the 'correct solution' should be.

**Comment by duncanmmacleod** [comment]
I would be happy for someone who understands the PyCBC build-and-test to take my commits and integrate them properly in a different PR.

**Comment by spxiwh** [comment]
I can take a look, but need to see what the failures actually are first.

**Comment by spxiwh** [comment]
I've proposed a PR onto this branch to get the test suite working .... It's basically what @duncanmmacleod was doing anyway.

@ahnitz Largely the problem is m2crypto, which does not provide pip wheels, so installing it through pip is a pain. It seems fine in conda, but needs swig + a few other things to work via pypi, as done in the Docker build (and of course mac also becomes a headache if trying to install m2crypto via pip). It's a bit annoying that m2crypto doesn't put proper wheels on pypi, but I think this should now get this unstuck.

This would need your approval before it can be merged, but I think (with the PR merged into it) that this would be good to go.

**Comment by spxiwh** [comment]
@ahnitz Can we split off this PR from the task of separating out the LVK-specific dependancies. That would probably need #4001 merged in any case. I think the basic test suite is already brining in dependencies it possibly doesn't need (openssl), so going through and checking that feels like a different task.

I promise to put that near the top of my to-do list if we can merge this as is!

**Comment by spxiwh** [comment]
AS the tests use conda, M2Crypto is build from a wheel with this change anyway, so it is not as flaky as if we were using `pip install`

**Comment by ahnitz** [comment]
@spxiwh I don't want to hold this up too much more, but I'd like clarification on your point about openssl. As far as I can tell. This was only added as a dependency in this PR, and I think only to support being able to install M2Crypto. Is that not needed?

**Review by spxiwh — APPROVED** [review]

**Pull request merged**
